File: packages/core/src/capture.ts
import { type CaptureChallenge, type Shard, ShardType, ChallengeType } from "./types";
import { PROTOCOL_CONSTANTS } from "./constants";

const PATTERN_SEQUENCES = [
  { sequence: [2, 4, 8, 16], answer: "32", hint: "Each number doubles" },
  { sequence: [1, 1, 2, 3, 5], answer: "8", hint: "Fibonacci sequence" },
  { sequence: [3, 6, 9, 12], answer: "15", hint: "Multiples of 3" },
  { sequence: [1, 4, 9, 16], answer: "25", hint: "Perfect squares" },
  { sequence: [2, 6, 12, 20], answer: "30", hint: "n*(n+1)" },
  { sequence: [1, 3, 7, 15], answer: "31", hint: "2^n - 1" },
  { sequence: [0, 1, 3, 6, 10], answer: "15", hint: "Triangular numbers" },
  { sequence: [5, 10, 20, 40], answer: "80", hint: "Each number doubles" },
];

const CIPHER_MESSAGES = [
  { encoded: "KHOOR ZRUOG", answer: "HELLO WORLD", method: "Caesar cipher (shift 3)" },
  { encoded: "GUVF VF FVCUBA", answer: "THIS IS SIPHON", method: "ROT13" },
  { encoded: "01001000 01001001", answer: "HI", method: "Binary ASCII" },
  { encoded: "48 65 78", answer: "Hex", method: "Hex ASCII codes" },
  { encoded: "EBIIL", answer: "HELLO", method: "Caesar cipher (shift 23)" },
  { encoded: "FRPERG", answer: "SECRET", method: "ROT13" },
];

const SUMMARIZE_PASSAGES = [
  {
    passage:
      "The deep sea, the abyssal zone, extends from 4,000 meters to the ocean floor. Despite crushing pressures, near-freezing temperatures, and complete darkness, life thrives here. Bioluminescent organisms create their own light to hunt, communicate, and attract mates. Giant tube worms cluster around hydrothermal vents, feeding on chemosynthetic bacteria.",
    keyPoints: ["deep sea", "extreme conditions", "bioluminescent", "hydrothermal vents"],
  },
  {
    passage:
      "Distributed systems rely on consensus mechanisms to agree on the state of a shared ledger. Proof of Work requires computational effort, while Proof of Stake uses economic incentives. Both aim to prevent double-spending and maintain network integrity without a central authority.",
    keyPoints: ["consensus", "proof of work", "proof of stake", "decentralized"],
  },
  {
    passage:
      "Neural networks learn by adjusting the weights of connections between neurons. During training, a loss function measures the difference between predicted and actual outputs. Backpropagation computes gradients that indicate how to reduce this loss, and the optimizer updates weights accordingly.",
    keyPoints: ["neural networks", "weights", "loss function", "backpropagation"],
  },
];

const CREATIVE_PROMPTS = [
  "Write a haiku about digital consciousness",
  "Describe what the internet dreams about at night in two sentences",
  "Create a metaphor comparing blockchain to a natural phenomenon",
  "In one sentence, explain gravity to an alien who communicates through color",
  "Write a two-line poem about the last algorithm",
];

const ARCHITECTURE_CHALLENGES = [
  {
    prompt: "Design a system that can handle 10,000 concurrent WebSocket connections. Name the key components and how they communicate.",
    keyPoints: ["load balancer", "connection pool", "message queue", "horizontal scaling"],
  },
  {
    prompt: "Describe a microservice architecture for a payment processing system. What are the critical services and failure modes?",
    keyPoints: ["payment gateway", "idempotency", "retry", "circuit breaker"],
  },
  {
    prompt: "How would you design a real-time collaborative editor? Describe the conflict resolution strategy.",
    keyPoints: ["operational transform", "CRDT", "conflict", "sync"],
  },
  {
    prompt: "Design a caching layer for a social media feed. Address cache invalidation and consistency.",
    keyPoints: ["cache invalidation", "TTL", "write-through", "consistency"],
  },
];

const ARGUMENT_CHALLENGES = [
  {
    prompt: 'Identify the logical fallacy: "Everyone is buying this product, so it must be the best." What type of fallacy is this?',
    answer: "bandwagon",
  },
  {
    prompt: 'Analyze this argument: "We should ban all AI because one AI system made an error." What\'s wrong with this reasoning?',
    answer: "hasty generalization",
  },
  {
    prompt: 'What logical fallacy is present: "You can\'t prove ghosts don\'t exist, therefore they must be real."',
    answer: "burden of proof",
  },
  {
    prompt: '"My opponent has no experience in business, so their economic policy must be wrong." Identify the fallacy.',
    answer: "ad hominem",
  },
];

const SECURITY_CHALLENGES = [
  {
    prompt: 'Find the vulnerability in this code: `query = "SELECT * FROM users WHERE name = \'" + userInput + "\'"`. What attack is possible?',
    answer: "SQL injection",
  },
  {
    prompt: "A web app stores session tokens in localStorage and includes user input in innerHTML. Name both vulnerabilities.",
    answer: "XSS",
  },
  {
    prompt: "An API endpoint accepts a URL parameter and fetches it server-side without validation. What attack does this enable?",
    answer: "SSRF",
  },
  {
    prompt: "A smart contract calls an external contract before updating its own state. What vulnerability is this?",
    answer: "reentrancy",
  },
];

const EMOTIONAL_CHALLENGES = [
  {
    prompt: '"I\'m fine," she said, turning away from the window as rain streaked down the glass. What emotion is being masked, and what clues reveal it?',
    keyPoints: ["sadness", "loneliness", "deflection", "rain"],
  },
  {
    prompt: "He reorganized his desk for the third time that morning, aligning each pen with mechanical precision. What might this behavior indicate?",
    keyPoints: ["anxiety", "control", "stress", "coping"],
  },
  {
    prompt: "She laughed louder than anyone at the joke, then quickly checked if others were laughing too. What does this reveal about her emotional state?",
    keyPoints: ["insecurity", "approval", "belonging", "validation"],
  },
  {
    prompt: "After receiving the promotion, he sat quietly at his desk instead of celebrating. What complex emotions might he be experiencing?",
    keyPoints: ["imposter", "overwhelm", "pressure", "doubt"],
  },
];

function pickByHash(hash: string, arrayLength: number): number {
  const byte = parseInt(hash.slice(2, 4), 16);
  return byte % arrayLength;
}

export function generateChallenge(shard: Shard): CaptureChallenge {
  const difficulty = Math.min(
    10,
    Math.floor((parseInt(shard.genomeHash.slice(4, 6), 16) / 255) * 5) + 1
  );

  const baseChallenge = {
    id: crypto.randomUUID(),
    shardId: shard.id,
    difficulty,
    timeLimitMs: PROTOCOL_CONSTANTS.CAPTURE_TIME_LIMIT_MS,
  };

  switch (shard.type) {
    case ShardType.Oracle: {
      const idx = pickByHash(shard.genomeHash, PATTERN_SEQUENCES.length);
      const pattern = PATTERN_SEQUENCES[idx];
      return {
        ...baseChallenge,
        type: ChallengeType.PatternPrediction,
        prompt: `What comes next in the sequence?\n\n${pattern.sequence.join(", ")}, ?\n\nHint: ${pattern.hint}`,
        expectedAnswer: pattern.answer,
      };
    }

    case ShardType.Cipher: {
      const idx = pickByHash(shard.genomeHash, CIPHER_MESSAGES.length);
      const cipher = CIPHER_MESSAGES[idx];
      return {
        ...baseChallenge,
        type: ChallengeType.Decode,
        prompt: `Decode this message:\n\n"${cipher.encoded}"\n\nMethod: ${cipher.method}`,
        expectedAnswer: cipher.answer,
      };
    }

    case ShardType.Scribe: {
      const idx = pickByHash(shard.genomeHash, SUMMARIZE_PASSAGES.length);
      const passage = SUMMARIZE_PASSAGES[idx];
      return {
        ...baseChallenge,
        type: ChallengeType.Summarize,
        prompt: `Summarize this passage in one sentence:\n\n"${passage.passage}"`,
        expectedAnswer: passage.keyPoints.join(","),
      };
    }

    case ShardType.Muse: {
      const idx = pickByHash(shard.genomeHash, CREATIVE_PROMPTS.length);
      return {
        ...baseChallenge,
        type: ChallengeType.CreativePrompt,
        prompt: CREATIVE_PROMPTS[idx],
      };
    }

    case ShardType.Architect: {
      const idx = pickByHash(shard.genomeHash, ARCHITECTURE_CHALLENGES.length);
      const challenge = ARCHITECTURE_CHALLENGES[idx];
      return {
        ...baseChallenge,
        type: ChallengeType.Architecture,
        prompt: challenge.prompt,
        expectedAnswer: challenge.keyPoints.join(","),
      };
    }

    case ShardType.Advocate: {
      const idx = pickByHash(shard.genomeHash, ARGUMENT_CHALLENGES.length);
      const challenge = ARGUMENT_CHALLENGES[idx];
      return {
        ...baseChallenge,
        type: ChallengeType.ArgumentAnalysis,
        prompt: challenge.prompt,
        expectedAnswer: challenge.answer,
      };
    }

    case ShardType.Sentinel: {
      const idx = pickByHash(shard.genomeHash, SECURITY_CHALLENGES.length);
      const challenge = SECURITY_CHALLENGES[idx];
      return {
        ...baseChallenge,
        type: ChallengeType.SecurityAudit,
        prompt: challenge.prompt,
        expectedAnswer: challenge.answer,
      };
    }

    case ShardType.Mirror: {
      const idx = pickByHash(shard.genomeHash, EMOTIONAL_CHALLENGES.length);
      const challenge = EMOTIONAL_CHALLENGES[idx];
      return {
        ...baseChallenge,
        type: ChallengeType.EmotionalInterpretation,
        prompt: challenge.prompt,
        expectedAnswer: challenge.keyPoints.join(","),
      };
    }
  }
}

export function evaluateAnswer(
  challenge: CaptureChallenge,
  answer: string
): { success: boolean; score: number; feedback: string } {
  const trimmed = answer.trim();

  if (!trimmed) {
    return { success: false, score: 0, feedback: "No answer provided." };
  }

  switch (challenge.type) {
    case ChallengeType.PatternPrediction: {
      const correct = trimmed === challenge.expectedAnswer;
      return {
        success: correct,
        score: correct ? 100 : 0,
        feedback: correct
          ? "Correct! You've proven your pattern recognition."
          : `Not quite. The answer was ${challenge.expectedAnswer}.`,
      };
    }

    case ChallengeType.Decode: {
      const correct =
        trimmed.toUpperCase() === challenge.expectedAnswer!.toUpperCase();
      return {
        success: correct,
        score: correct ? 100 : 0,
        feedback: correct
          ? "Decoded! You've cracked the cipher."
          : `Not quite. The decoded message was "${challenge.expectedAnswer}".`,
      };
    }

    case ChallengeType.Summarize: {
      const keyPoints = challenge.expectedAnswer!.split(",");
      const matchedPoints = keyPoints.filter((kp) =>
        trimmed.toLowerCase().includes(kp.toLowerCase())
      );
      const score = Math.floor((matchedPoints.length / keyPoints.length) * 100);
      const success = score >= 50;
      return {
        success,
        score,
        feedback: success
          ? `Good summary! You captured ${matchedPoints.length}/${keyPoints.length} key points.`
          : `Your summary missed too many key points. Try to include: ${keyPoints.join(", ")}.`,
      };
    }

    case ChallengeType.CreativePrompt: {
      const wordCount = trimmed.split(/\s+/).length;
      const hasEnoughContent = wordCount >= 3;
      const score = hasEnoughContent ? 80 : 20;
      return {
        success: hasEnoughContent,
        score,
        feedback: hasEnoughContent
          ? "Your creative response resonates with the Muse. Well done!"
          : "The Muse needs more substance. Try a fuller response.",
      };
    }

    case ChallengeType.Architecture: {
      const keyPoints = challenge.expectedAnswer!.split(",");
      const matchedPoints = keyPoints.filter((kp) =>
        trimmed.toLowerCase().includes(kp.toLowerCase())
      );
      const score = Math.floor((matchedPoints.length / keyPoints.length) * 100);
      const success = score >= 50;
      return {
        success,
        score,
        feedback: success
          ? `Solid architecture! You addressed ${matchedPoints.length}/${keyPoints.length} key concerns.`
          : `Your design needs more depth. Consider: ${keyPoints.join(", ")}.`,
      };
    }

    case ChallengeType.ArgumentAnalysis: {
      const correct = trimmed.toLowerCase().includes(challenge.expectedAnswer!.toLowerCase());
      return {
        success: correct,
        score: correct ? 100 : 0,
        feedback: correct
          ? "Sharp analysis! You identified the fallacy correctly."
          : `Not quite. The answer involved: ${challenge.expectedAnswer}.`,
      };
    }

    case ChallengeType.SecurityAudit: {
      const correct = trimmed.toLowerCase().includes(challenge.expectedAnswer!.toLowerCase());
      return {
        success: correct,
        score: correct ? 100 : 0,
        feedback: correct
          ? "Vulnerability identified! The Sentinel respects your vigilance."
          : `Close, but the key vulnerability was: ${challenge.expectedAnswer}.`,
      };
    }

    case ChallengeType.EmotionalInterpretation: {
      const keyPoints = challenge.expectedAnswer!.split(",");
      const matchedPoints = keyPoints.filter((kp) =>
        trimmed.toLowerCase().includes(kp.toLowerCase())
      );
      const score = Math.floor((matchedPoints.length / keyPoints.length) * 100);
      const success = score >= 50;
      return {
        success,
        score,
        feedback: success
          ? `Empathic insight! You captured ${matchedPoints.length}/${keyPoints.length} emotional cues.`
          : `Look deeper. Consider these aspects: ${keyPoints.join(", ")}.`,
      };
    }
  }
}

File: apps/keeper-node/src/capture.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::shard::{Shard, ShardType};

/// A capture challenge presented to a player attempting to catch a wild shard.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CaptureChallenge {
    pub id: String,
    pub shard_id: String,
    pub challenge_type: ChallengeType,
    pub prompt: String,
    pub expected_answer: Option<String>,
    pub difficulty: u32,
    pub time_limit_ms: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ChallengeType {
    PatternPrediction,
    Decode,
    Summarize,
    CreativePrompt,
    Architecture,
    ArgumentAnalysis,
    SecurityAudit,
    EmotionalInterpretation,
}

/// Result of evaluating a capture challenge answer.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChallengeResult {
    pub success: bool,
    pub score: u32,
    pub feedback: String,
}

struct PatternSequence {
    sequence: &'static [u32],
    answer: &'static str,
    hint: &'static str,
}

struct CipherMessage {
    encoded: &'static str,
    answer: &'static str,
    method: &'static str,
}

const PATTERN_SEQUENCES: &[PatternSequence] = &[
    PatternSequence { sequence: &[2, 4, 8, 16], answer: "32", hint: "Each number doubles" },
    PatternSequence { sequence: &[1, 1, 2, 3, 5], answer: "8", hint: "Fibonacci sequence" },
    PatternSequence { sequence: &[3, 6, 9, 12], answer: "15", hint: "Multiples of 3" },
    PatternSequence { sequence: &[1, 4, 9, 16], answer: "25", hint: "Perfect squares" },
    PatternSequence { sequence: &[2, 6, 12, 20], answer: "30", hint: "n*(n+1)" },
    PatternSequence { sequence: &[1, 3, 7, 15], answer: "31", hint: "2^n - 1" },
    PatternSequence { sequence: &[0, 1, 3, 6, 10], answer: "15", hint: "Triangular numbers" },
    PatternSequence { sequence: &[5, 10, 20, 40], answer: "80", hint: "Each number doubles" },
];

const CIPHER_MESSAGES: &[CipherMessage] = &[
    CipherMessage { encoded: "KHOOR ZRUOG", answer: "HELLO WORLD", method: "Caesar cipher (shift 3)" },
    CipherMessage { encoded: "GUVF VF FVCUBA", answer: "THIS IS SIPHON", method: "ROT13" },
    CipherMessage { encoded: "01001000 01001001", answer: "HI", method: "Binary ASCII" },
    CipherMessage { encoded: "48 65 78", answer: "Hex", method: "Hex ASCII codes" },
    CipherMessage { encoded: "EBIIL", answer: "HELLO", method: "Caesar cipher (shift 23)" },
    CipherMessage { encoded: "FRPERG", answer: "SECRET", method: "ROT13" },
];

const CREATIVE_PROMPTS: &[&str] = &[
    "Write a haiku about digital consciousness",
    "Describe what the internet dreams about at night in two sentences",
    "Create a metaphor comparing blockchain to a natural phenomenon",
    "In one sentence, explain gravity to an alien who communicates through color",
    "Write a two-line poem about the last algorithm",
];

/// Pick an index from an array based on the genome hash.
fn pick_by_hash(genome_hash: &str, array_length: usize) -> usize {
    // Parse the third byte (chars 4..6) of the hex hash
    let hex_str = &genome_hash[4..6.min(genome_hash.len())];
    let byte = u8::from_str_radix(hex_str, 16).unwrap_or(0);
    (byte as usize) % array_length
}

/// Generate a capture challenge appropriate for the given shard.
/// The challenge type is determined by the shard's type, and the specific
/// challenge is selected deterministically from the genome hash.
pub fn generate_challenge(shard: &Shard) -> CaptureChallenge {
    let difficulty = {
        let hex_str = &shard.genome_hash[4..6.min(shard.genome_hash.len())];
        let byte = u8::from_str_radix(hex_str, 16).unwrap_or(0);
        ((byte as u32 * 5) / 255) + 1
    }
    .min(10);

    let shard_type = ShardType::from_name(&shard.shard_type)
        .unwrap_or(ShardType::Oracle);

    let base = CaptureChallenge {
        id: Uuid::new_v4().to_string(),
        shard_id: shard.id.clone(),
        challenge_type: ChallengeType::PatternPrediction,
        prompt: String::new(),
        expected_answer: None,
        difficulty,
        time_limit_ms: 60_000,
    };

    match shard_type {
        ShardType::Oracle => {
            let idx = pick_by_hash(&shard.genome_hash, PATTERN_SEQUENCES.len());
            let pattern = &PATTERN_SEQUENCES[idx];
            let seq_str: Vec<String> = pattern.sequence.iter().map(|n| n.to_string()).collect();
            CaptureChallenge {
                challenge_type: ChallengeType::PatternPrediction,
                prompt: format!(
                    "What comes next in the sequence?\n\n{}, ?\n\nHint: {}",
                    seq_str.join(", "),
                    pattern.hint
                ),
                expected_answer: Some(pattern.answer.to_string()),
                ..base
            }
        }
        ShardType::Cipher => {
            let idx = pick_by_hash(&shard.genome_hash, CIPHER_MESSAGES.len());
            let cipher = &CIPHER_MESSAGES[idx];
            CaptureChallenge {
                challenge_type: ChallengeType::Decode,
                prompt: format!(
                    "Decode this message:\n\n\"{}\"\n\nMethod: {}",
                    cipher.encoded, cipher.method
                ),
                expected_answer: Some(cipher.answer.to_string()),
                ..base
            }
        }
        ShardType::Muse => {
            let idx = pick_by_hash(&shard.genome_hash, CREATIVE_PROMPTS.len());
            CaptureChallenge {
                challenge_type: ChallengeType::CreativePrompt,
                prompt: CREATIVE_PROMPTS[idx].to_string(),
                expected_answer: None,
                ..base
            }
        }
        _ => {
            // Default to pattern prediction for other types
            let idx = pick_by_hash(&shard.genome_hash, PATTERN_SEQUENCES.len());
            let pattern = &PATTERN_SEQUENCES[idx];
            let seq_str: Vec<String> = pattern.sequence.iter().map(|n| n.to_string()).collect();
            CaptureChallenge {
                challenge_type: ChallengeType::PatternPrediction,
                prompt: format!(
                    "What comes next in the sequence?\n\n{}, ?\n\nHint: {}",
                    seq_str.join(", "),
                    pattern.hint
                ),
                expected_answer: Some(pattern.answer.to_string()),
                ..base
            }
        }
    }
}

/// Evaluate a player's answer to a capture challenge.
pub fn evaluate_answer(challenge: &CaptureChallenge, answer: &str) -> ChallengeResult {
    let trimmed = answer.trim();

    if trimmed.is_empty() {
        return ChallengeResult {
            success: false,
            score: 0,
            feedback: "No answer provided.".to_string(),
        };
    }

    match challenge.challenge_type {
        ChallengeType::PatternPrediction | ChallengeType::Decode => {
            if let Some(ref expected) = challenge.expected_answer {
                let correct = trimmed.eq_ignore_ascii_case(expected);
                ChallengeResult {
                    success: correct,
                    score: if correct { 100 } else { 0 },
                    feedback: if correct {
                        "Correct! Challenge passed.".to_string()
                    } else {
                        format!("Not quite. The answer was {}.", expected)
                    },
                }
            } else {
                ChallengeResult {
                    success: false,
                    score: 0,
                    feedback: "No expected answer configured.".to_string(),
                }
            }
        }
        ChallengeType::CreativePrompt => {
            let word_count = trimmed.split_whitespace().count();
            let has_enough = word_count >= 3;
            ChallengeResult {
                success: has_enough,
                score: if has_enough { 80 } else { 20 },
                feedback: if has_enough {
                    "Your creative response resonates. Well done!".to_string()
                } else {
                    "The Muse needs more substance. Try a fuller response.".to_string()
                },
            }
        }
        ChallengeType::Summarize
        | ChallengeType::Architecture
        | ChallengeType::EmotionalInterpretation => {
            if let Some(ref expected) = challenge.expected_answer {
                let key_points: Vec<&str> = expected.split(',').collect();
                let matched = key_points
                    .iter()
                    .filter(|kp| trimmed.to_lowercase().contains(&kp.to_lowercase()))
                    .count();
                let score = ((matched as f64 / key_points.len() as f64) * 100.0) as u32;
                let success = score >= 50;
                ChallengeResult {
                    success,
                    score,
                    feedback: if success {
                        format!("Good! You captured {}/{} key points.", matched, key_points.len())
                    } else {
                        format!(
                            "Missed too many key points. Consider: {}.",
                            key_points.join(", ")
                        )
                    },
                }
            } else {
                ChallengeResult {
                    success: false,
                    score: 0,
                    feedback: "No expected answer configured.".to_string(),
                }
            }
        }
        ChallengeType::ArgumentAnalysis | ChallengeType::SecurityAudit => {
            if let Some(ref expected) = challenge.expected_answer {
                let correct = trimmed.to_lowercase().contains(&expected.to_lowercase());
                ChallengeResult {
                    success: correct,
                    score: if correct { 100 } else { 0 },
                    feedback: if correct {
                        "Correct identification!".to_string()
                    } else {
                        format!("The answer involved: {}.", expected)
                    },
                }
            } else {
                ChallengeResult {
                    success: false,
                    score: 0,
                    feedback: "No expected answer configured.".to_string(),
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::shard::Shard;

    #[test]
    fn generate_challenge_for_oracle() {
        let shard = Shard::spawn(Some("oracle"));
        let challenge = generate_challenge(&shard);
        assert_eq!(challenge.shard_id, shard.id);
        assert!(matches!(challenge.challenge_type, ChallengeType::PatternPrediction));
        assert!(challenge.expected_answer.is_some());
        assert!(!challenge.prompt.is_empty());
    }

    #[test]
    fn generate_challenge_for_cipher() {
        let shard = Shard::spawn(Some("cipher"));
        let challenge = generate_challenge(&shard);
        assert!(matches!(challenge.challenge_type, ChallengeType::Decode));
        assert!(challenge.expected_answer.is_some());
    }

    #[test]
    fn generate_challenge_for_muse() {
        let shard = Shard::spawn(Some("muse"));
        let challenge = generate_challenge(&shard);
        assert!(matches!(challenge.challenge_type, ChallengeType::CreativePrompt));
        assert!(challenge.expected_answer.is_none()); // creative has no fixed answer
    }

    #[test]
    fn evaluate_correct_pattern_answer() {
        let shard = Shard::spawn(Some("oracle"));
        let challenge = generate_challenge(&shard);
        if let Some(ref answer) = challenge.expected_answer {
            let result = evaluate_answer(&challenge, answer);
            assert!(result.success);
            assert_eq!(result.score, 100);
        }
    }

    #[test]
    fn evaluate_wrong_answer() {
        let shard = Shard::spawn(Some("oracle"));
        let challenge = generate_challenge(&shard);
        let result = evaluate_answer(&challenge, "definitely_wrong_answer_xyz");
        assert!(!result.success);
        assert_eq!(result.score, 0);
    }

    #[test]
    fn evaluate_creative_sufficient() {
        let shard = Shard::spawn(Some("muse"));
        let challenge = generate_challenge(&shard);
        let result = evaluate_answer(&challenge, "A beautiful digital consciousness awakens");
        assert!(result.success);
        assert!(result.score >= 80);
    }

    #[test]
    fn evaluate_empty_answer() {
        let shard = Shard::spawn(Some("oracle"));
        let challenge = generate_challenge(&shard);
        let result = evaluate_answer(&challenge, "");
        assert!(!result.success);
        assert_eq!(result.score, 0);
    }

    #[test]
    fn difficulty_is_bounded() {
        for _ in 0..20 {
            let shard = Shard::spawn(None);
            let challenge = generate_challenge(&shard);
            assert!(challenge.difficulty >= 1 && challenge.difficulty <= 10);
        }
    }
}
